### ISSUE TITLE
Some inconsistencies when using Or matching

### ISSUE THREAD

**Issue by KristofferC** [FluxML/MacroTools.jl#193]
Typically, if `@capture(expr, match_1)` matches you would expect `@capture(expr, match_1 | match_2)` to always also match. This is however not the case. For example:

```julia
julia> @capture(:(begin; x = 2;y=3; end), begin body_ end)
true

julia> body
quote
    #= REPL[2]:1 =#
    x = 2
    #= REPL[2]:1 =#
    y = 3
end
```

matches, but the following

```julia
julia> @capture(:(begin; x = 2;y=3; end), begin body_ end | for i_ in iter_ forbody_ end)
false
```

does not. In addition, it the expression in the block only contains one expression, there will be a match:

```julia
julia> @capture(:(begin; x = 2; end), begin body_ end | for i_ in iter_ forbody_ end)
true

julia> body
:(x = 2)
```

This feels inconsistent to me.

**Comment by cstjean** [comment]
Agreed that it's inconsistent. Not sure which way we should resolve it though. 
```julia
julia> @capture(:(begin; x = 2;y=3; end), begin body_ end)
true
```
felt surprising to me, but at the same time I can kinda see the reasoning behind it. I'm a bit worried that if we changed that to false, we'd break people's code somewhere, and might learn the original reason for this behaviour.

So that leaves us with making

```julia
julia> @capture(:(begin; x = 2;y=3; end), begin body_ end | for i_ in iter_ forbody_ end)
false
```
true. Not a great fan, but it's probably safer.

**Comment by heyx3** [comment]
Isn't this just a straightforward bug? I can't see why the second statement should ever fail to match.

Plus Julia's versioning system allows a for package to indicate that it has breaking changes.

**Comment by heyx3** [comment]
In fact I just noticed this trivial example returns false:

`@capture 4 (x_Int | x_String)`

It seems like `|` is totally broken unless I'm missing something?

**Comment by cstjean** [comment]
I would welcome a PR that makes the second one true, for consideration. I agree that it's basically a bugfix.

I haven't looked at that code, but I suspect the issue is that `begin x end` is special-cased to be equivalent to `x` in the name of convenience (very rarely would one want to distinguish the two), but that special-casing isn't uniformly applied. 

> It seems like | is totally broken unless I'm missing something?

I rarely use `|` and the `_type` forms, but indeed that looks broken.
